### PR TITLE
BUG: Fix corner-case behavior of cond() and use SVD when possible

### LINKED ISSUE #7907: Could `np.linalg.cond` make a more effective use of SVD computations?
The condition number is defined as `norm(A) * norm(inv(A))`. In the case of `np.linalg.cond` when kwarg `p` is set to `+/-2` calls to `norm(A, p=+/-2)` and `norm(inv(A), p=+/2)` are made. This is likely inefficient and possibly numerically weak due to the following:

In the cases of `p == 2` or `p ==-2` singular values are used for computing norms, these coming from the singular value decomposition of `A`, a costly operation. Recall that the singular values of an inverted matrix are the reciprocal of singular values of the original matrix. Therefore calling `svd(A)` once yields all the information needed about both `A` and `inv(A)` without the cost or potential loss of accuracy incurred through inversion and a second call to `svd`. For the case of `p == 2`, the result is just the ratio of `largest singular value/smallest singular value`, and for the case of `p==-2` the result is simply the `smallest singular value/largest singular value`. This is implemented as described, in numba, here:
https://github.com/numba/numba/blob/master/numba/targets/linalg.py#L2072-L2101

The numpy code in question is:
https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py#L1459

If there is agreement over this I'm happy to take a shot at fixing it.

### LINKED ISSUE #8090: linalg.cond does not check for exact zero
For the following case, the convention is to give `np.inf` which is also consistent with the documentation

```
C = np.zeros((3,3))
np.linalg.cond(C)
```

however, does the following 

```
c:\....\appdata\local\programs\python\python35\lib\site-packages\numpy\linalg\linalg.py:1457: 
RuntimeWarning: invalid value encountered in true_divide
  return s[..., 0]/s[..., -1]
Out[1]: nan
```

Similarly, if a singular matrix is given, the algorithm does not check the health of the inversion but direclty inverts the matrix regardless of the norm used.  Hence, 

```
cond([[1,1],[1,1]],2)
```

thus throws a `LinAlgError` for singular matrix. This defeats the purpose of the function since I am trying to monitor this information without the exception. I guess a simple check for exact zero singular values would rule out these possibilities.

### PR DESCRIPTION
**Pull request by pv**
Make np.linalg.cond(A, p) always use SVD for p=+-2 and not only p=None.

Ensure inf is returned instead of nan when the condition number diverges
for +/-2 norm.

Ignore errors when inverting matrices for p = +-1, 'fro', +-inf, so that
spurious LinAlgErrors are not raised.  In those cases, report
non-invertible matrices as having cond=inf.

For this latter set of norms, the computation is still somewhat naively done because we are not basing it on a decomposition, so the number computed when the matrix is close to singular will be large but not very accurate. On the other hand, exact and stable computation of the condition numbers in these norms probably requires quite a bit more care, and svd/qr/etc are not as immediately helpful as for 2-norm. (I think LAPACK has routines that estimate the condition number in 1-norm, but no exact computation.)

Add additional tests for cond().

Fixes gh-8090
Fixes gh-7907

### CODE REVIEW

File: numpy/linalg/linalg.py [hunk 1]
```diff
@@ -1538,11 +1538,34 @@ def cond(x, p=None):
...
+            r = norm(x, p, axis=(-2, -1)) * norm(invx, p, axis=(-2, -1))
+        r = r.astype(result_t, copy=False)
+
+    if r.ndim > 0:
+        r[~isfinite(r)] = Inf
+    elif not isfinite(r):
+        if isinstance(r, ndarray):
+            r[()] = Inf
+        else:
+            r = type(r)(Inf)
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: How about:

```python
r = asarray(r)
r[~isfinite(r)] = Inf
return r[()]
```

Comment (pv) [reply numpy/linalg/linalg.py]: Ok, changed. Also fixed handling of nans in inputs.

**Comment by charris** [comment]
Thanks Pauli.

File: numpy/linalg/linalg.py [hunk 2]
```diff
@@ -1538,11 +1538,40 @@ def cond(x, p=None):
...
+
+    # Convert nans to infs unless the original array had nan entries
+    r = asarray(r)
+    nan_mask = isnan(r)
+    if nan_mask.any():
+        nan_mask &= ~isnan(x).any(axis=(-2, -1))
+        if r.ndim > 0:
+            r[nan_mask] = Inf
+        elif nan_mask:
+            r[()] = Inf
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: You don't need this special case, do you? Assigning a mask works just fine for 0d as well

**Comment by pv** [comment]
@eric-wieser: correct, the special case should be possible to remove.

**Pull request merged**